### ISSUE TITLE
Questions about the performance

### ISSUE THREAD

**Issue by chavinlo** [facebookresearch/CrypTen#423]
Hello, I was planning on using CrypTen to finetune a SD model with hivemind but from what I was able to read theres a decrease in performance thats "100x" slower than normal PyTorch on CPU and "10x" slower on GPU (src.: Paper pg. 7 , #414 , #360 )
Hivemind shares the gradients and parameters with multiple nodes and averages them (from what I've read).

Is there a way to optimize it? I only need to make it possible to train on top of it but not to do inference with it, I guess some sort of passphrase to decrypt it for inference. The dataset is public so it can be unencrypted but I don't think thats the main source of the performance decrease right? And if it's not possible, are there any viable alternatives?

**Comment by mohammad-alrubaie** [comment]
It is expected that any SMPC library, including CrypTen, would have lower performance than a library doing computation on without security/privacy guarantees. Furthermore, whenever you have a fixed set of computations, you can highly optimize certain components, and achieve higher performance. CrypTen, on the other hand, is very flexible, and it lowers the barrier to entry for ML engineers. If you see the paper, you will also see that different architectures were used in experiments, and that revealed certain bottlenecks for certain architectures. So, it would be great to use CrypTen for a general-purpose use case, and then identify what are the bottlenecks for your specific architecture, to optimize those, but that may require a specialized protocol. 
Also, from your description, it seems that the dataset is public, and inference can also be public. What is the threat model you're dealing with?
